File: kernel_design/executable_spec/assembly/result_routing_application.py
from __future__ import annotations

import logging
from collections.abc import Callable
from dataclasses import dataclass
from functools import partial
from threading import Event, Lock, Thread

from ..result_routing import ResultRoutingConfig, ResultRoutingPacer


_LOGGER = logging.getLogger(__name__)


@dataclass(frozen=True, slots=True)
class ResultRoutingApplicationConfig:
    routing: ResultRoutingConfig
    interval_millis: int

    def __post_init__(self) -> None:
        if self.interval_millis <= 0:
            raise ValueError("result-routing interval must be positive")


class ResultRoutingApplication:
    """Application lifecycle for the bounded result-routing loop."""

    def __init__(self, pacer: ResultRoutingPacer) -> None:
        self.pacer = pacer
        self._lifecycle_lock = Lock()
        self._stop_event: Event | None = None
        self._thread: Thread | None = None

    def start(self, *, config: ResultRoutingApplicationConfig) -> None:
        with self._lifecycle_lock:
            if self._thread is not None:
                raise RuntimeError("result-routing application is already started")

            stop_event = Event()
            thread = Thread(
                name="result-routing",
                target=self._run_loop,
                kwargs={
                    "stop_event": stop_event,
                    "interval_millis": config.interval_millis,
                    "operation": partial(
                        self.pacer.route_seed_results,
                        config=config.routing,
                    ),
                },
                daemon=False,
            )
            thread.start()
            self._stop_event = stop_event
            self._thread = thread

    def stop(self, *, timeout_millis: int = 5_000) -> None:
        if timeout_millis <= 0:
            raise ValueError("stop timeout must be positive")

        with self._lifecycle_lock:
            thread = self._thread
            stop_event = self._stop_event
            if thread is None or stop_event is None:
                return
            stop_event.set()

        thread.join(timeout=timeout_millis / 1_000)
        if thread.is_alive():
            raise TimeoutError("result-routing loop did not stop")

        with self._lifecycle_lock:
            if self._thread is thread:
                self._thread = None
                self._stop_event = None

    @staticmethod
    def _run_loop(
        *,
        stop_event: Event,
        interval_millis: int,
        operation: Callable[[], int],
    ) -> None:
        while not stop_event.is_set():
            try:
                operation()
            except Exception:
                _LOGGER.exception("result-routing round failed")
            if stop_event.wait(interval_millis / 1_000):
                return

File: kernel_jvm/src/main/java/com/xa/mass/kernel/result/ResultRoutingApplication.java
package com.xa.mass.kernel.result;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

public final class ResultRoutingApplication {

    private static final System.Logger LOGGER = System.getLogger(
            ResultRoutingApplication.class.getName()
    );

    private final Object lifecycleLock = new Object();
    private final ResultRoutingPacer pacer;
    private Thread thread;
    private CountDownLatch stopSignal;
    private State state = State.STOPPED;

    public ResultRoutingApplication(ResultRoutingPacer pacer) {
        this.pacer = java.util.Objects.requireNonNull(pacer, "pacer");
    }

    public void start(ResultRoutingApplicationConfig config) {
        java.util.Objects.requireNonNull(config, "config");
        synchronized (lifecycleLock) {
            if (thread != null || state != State.STOPPED) {
                throw new IllegalStateException(
                        "Result Routing application is already started"
                );
            }
            CountDownLatch signal = new CountDownLatch(1);
            Thread started = new Thread(
                    () -> runLoop(config, signal),
                    "result-routing"
            );
            started.setDaemon(false);
            stopSignal = signal;
            thread = started;
            state = State.RUNNING;
            started.start();
        }
    }

    public void stop(long timeoutMillis) {
        if (timeoutMillis < 1) {
            throw new IllegalArgumentException(
                    "timeoutMillis must be positive"
            );
        }
        Thread current;
        CountDownLatch signal;
        synchronized (lifecycleLock) {
            current = thread;
            signal = stopSignal;
            if (current == null || signal == null) {
                return;
            }
            state = State.STOPPING;
            signal.countDown();
        }
        try {
            current.join(timeoutMillis);
        } catch (InterruptedException error) {
            Thread.currentThread().interrupt();
            throw new IllegalStateException(
                    "Result Routing shutdown was interrupted",
                    error
            );
        }
        if (current.isAlive()) {
            throw new IllegalStateException(
                    "Result Routing did not stop within its budget"
            );
        }
        synchronized (lifecycleLock) {
            if (thread == current) {
                thread = null;
                stopSignal = null;
                state = State.STOPPED;
            }
        }
    }

    public boolean isRunning() {
        synchronized (lifecycleLock) {
            return state == State.RUNNING
                    && thread != null
                    && thread.isAlive();
        }
    }

    public String state() {
        synchronized (lifecycleLock) {
            refreshDeadThread();
            return state.name();
        }
    }

    private void runLoop(
            ResultRoutingApplicationConfig config,
            CountDownLatch signal
    ) {
        try {
            while (signal.getCount() > 0) {
                try {
                    pacer.routeWorkerResults(config.routing());
                } catch (RuntimeException error) {
                    LOGGER.log(
                            System.Logger.Level.ERROR,
                            "operation=resultRouting.routeWorkerResults "
                                    + "failed",
                            error
                    );
                }
                if (signal.await(
                        config.intervalMillis(),
                        TimeUnit.MILLISECONDS
                )) {
                    return;
                }
            }
        } catch (InterruptedException error) {
            Thread.currentThread().interrupt();
        } finally {
            synchronized (lifecycleLock) {
                if (thread == Thread.currentThread()
                        && state != State.STOPPING) {
                    state = State.FAILED;
                }
            }
        }
    }

    private void refreshDeadThread() {
        if (state == State.RUNNING
                && thread != null
                && !thread.isAlive()) {
            state = State.FAILED;
        }
    }

    private enum State {
        STOPPED,
        RUNNING,
        STOPPING,
        FAILED
    }
}
